### ISSUE TITLE
yolov5做nms时，有一个agnostic参数，确定是否在类间做nms，但fastdeploy的predict函数只有nms的iou阈值设置，不知道这个是在哪块设置呢？

### ISSUE THREAD

**Issue by liuyiche** [PaddlePaddle/FastDeploy#887]

**Comment by wjj19950828** [comment]
@liuyiche 请问下需要这个参数具体需求是什么呢，这个参数会影响到最终ap值或者结果吗

**Comment by liuyiche** [comment]
> @liuyiche 请问下需要这个参数具体需求是什么呢，这个参数会影响到最终ap值或者结果吗

假如模型输出有两个框的iou超过nms阈值，但是类别不一样，我这边需要过滤下，只留下一个置信度高的，原本的yolov5会有这个参数，就是用来过滤的。。    会影响ap值和结果的。

**Comment by liuyiche** [comment]
> @liuyiche 请问下需要这个参数具体需求是什么呢，这个参数会影响到最终ap值或者结果吗

def non_max_suppression(prediction, conf_thres=0.25, iou_thres=0.45, classes=None, agnostic=False, multi_label=False,
                        labels=(), max_det=300):
    """Runs Non-Maximum Suppression (NMS) on inference results

    Returns:
         list of detections, on (n,6) tensor per image [xyxy, conf, cls]
    """

    nc = prediction.shape[2] - 5  # number of classes
    xc = prediction[..., 4] > conf_thres  # candidates

    # Checks
    assert 0 <= conf_thres <= 1, f'Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0'
    assert 0 <= iou_thres <= 1, f'Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0'

    # Settings
    min_wh, max_wh = 2, 4096  # (pixels) minimum and maximum box width and height
    max_nms = 30000  # maximum number of boxes into torchvision.ops.nms()
    time_limit = 10.0  # seconds to quit after
    redundant = True  # require redundant detections
    multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)
    merge = False  # use merge-NMS

    t = time.time()
    output = [torch.zeros((0, 6), device=prediction.device)] * prediction.shape[0]
    for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        # x[((x[..., 2:4] < min_wh) | (x[..., 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  # confidence

        # Cat apriori labels if autolabelling
        if labels and len(labels[xi]):
            l = labels[xi]
            v = torch.zeros((len(l), nc + 5), device=x.device)
            v[:, :4] = l[:, 1:5]  # box
            v[:, 4] = 1.0  # conf
            v[range(len(l)), l[:, 0].long() + 5] = 1.0  # cls
            x = torch.cat((x, v), 0)

        # If none remain process next image
        if not x.shape[0]:
            continue

        # Compute conf
        x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf

        # Box (center x, center y, width, height) to (x1, y1, x2, y2)

        box = xywh2xyxy(x[:, :4])

        # Detections matrix nx6 (xyxy, conf, cls)
        if multi_label:
            i, j = (x[:, 5:] > conf_thres).nonzero(as_tuple=False).T
            x = torch.cat((box[i], x[i, j + 5, None], j[:, None].float()), 1)
        else:  # best class only
            conf, j = x[:, 5:].max(1, keepdim=True)
            x = torch.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres]

        # Filter by class
        if classes is not None:
            x = x[(x[:, 5:6] == torch.tensor(classes, device=x.device)).any(1)]

        # Apply finite constraint
        # if not torch.isfinite(x).all():
        #     x = x[torch.isfinite(x).all(1)]

        # Check shape

        n = x.shape[0]  # number of boxes
        if not n:  # no boxes
            continue
        elif n > max_nms:  # excess boxes

            x = x[x[:, 4].argsort(descending=True)[:max_nms]]  # sort by confidence

        # Batched NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        if i.shape[0] > max_det:  # limit detections
            i = i[:max_det]
        if merge and (1 < n < 3E3):  # Merge NMS (boxes merged using weighted mean)
            # update boxes as boxes(i,4) = weights(i,n) * boxes(n,4)
            iou = box_iou(boxes[i], boxes) > iou_thres  # iou matrix
            weights = iou * scores[None]  # box weights
            x[i, :4] = torch.mm(weights, x[:, :4]).float() / weights.sum(1, keepdim=True)  # merged boxes
            if redundant:
                i = i[iou.sum(1) > 1]  # require redundancy

        output[xi] = x[i]
        if (time.time() - t) > time_limit:
            logger.info(f'WARNING: NMS time limit {time_limit}s exceeded')
            break  # time limit exceeded

    return output





这个是yolov5的nms函数，agnostic参数设置True或者False

**Comment by wjj19950828** [comment]
@liuyiche 麻烦提供一下这个参数影响的图片吧，我这边测试下

我理解这个参数是YOLOv5专门为多类别设置的参数，默认为Fasle，设置为True后在多分类任务中AP会下降，参考https://github.com/ultralytics/yolov5/issues/621

官方在detect.py以及eval.py中默认值均为False

**Comment by liuyiche** [comment]
> 

嗯，默认确实为False，一般影响不大，或者再做一次处理就行，主要是对一些难样本可能会出现预测的类别重叠导致需要过滤。   主要是原版yolov5自带的，咱们后期更新能加进来更好些。因为不是在公开数据集训练的，测试需要同时提供图片和模型，稍有不便。

**Comment by liuyiche** [comment]
> 

嗯，默认确实为False，一般影响不大，或者再做一次处理就行，主要是对一些难样本可能会出现预测的类别重叠导致需要过滤。   主要是原版yolov5自带的，咱们后期更新能加进来更好些。因为不是在公开数据集训练的，测试需要同时提供图片和模型，稍有不便。

**Comment by jiangjiajun** [comment]
此ISSUE由于一年未更新，将会关闭处理，如有需要，可再次更新打开。
